### PR TITLE
[Web]-Etp 123 web bug task update fails with zod validation error leading to foreign key constraint violation

### PR DESCRIPTION
**Pull request by NdekoCode**
# 🚀 Fix Task Update Failures with Mixed Member Structures and Zod Validation

_Resolves Zod validation errors and foreign key constraint violations when updating tasks with mixed member structures._

## Description

This PR fixes critical task update failures caused by mixed member structures in `task.members` arrays. The issue occurred when tasks contained both Direct Employee and OrganizationTeamEmployee structures, leading to Zod validation failures and subsequent foreign key constraint violations.

**Problems Addressed:**
- Zod validation error: `members.1.employee.email: Required`
- Foreign key constraint violation: `insert or update on table "task_employee" violates foreign key constraint "FK_f38b1bd46f8831704348003bbff"`
- Task assignment functionality completely broken
- User workflow disruption in task management

**Root Cause:**
Tasks can contain mixed member structures where some members use `userId` (Direct Employee) and others use `employeeId` (OrganizationTeamEmployee). The API backend expects a uniform structure based on `userId`, but the frontend was sending inconsistent data structures.

**Solution:**
Implemented member structure normalization that converts all members to a consistent `userId`-based format before sending to the API, while preserving all necessary data and maintaining backward compatibility.

### Major Changes

- **Added Member Structure Normalization**: Created intelligent normalization logic in `TaskService.updateTask()` that handles mixed member structures
- **Refactored Task Service**: Extracted normalization logic into clean, testable private methods:
  - `getMemberType()`: Determines member type (DirectEmployee vs OrganizationTeamEmployee)
  - `normalizeDirectEmployee()`: Handles Direct Employee normalization
  - `normalizeOrganizationTeamEmployee()`: Converts OrganizationTeamEmployee to userId-based structure
  - `normalizeTaskMembers()`: Orchestrates the complete normalization process
- **Re-enabled Zod Validation**: Restored client-side validation that now works with normalized data structures
- **Enhanced Error Handling**: Added comprehensive logging and error handling throughout the normalization process

### Minor Changes

- **Improved Code Readability**: Replaced 45+ lines of nested if/else statements with clean, functional programming approach
- **Added Detailed Logging**: Comprehensive console logging for debugging and monitoring
- **Type Safety**: Improved TypeScript usage with proper type annotations

## How to Test This PR

Please follow these steps to test the changes locally:

1. **Setup and Run**:
   ```bash
   yarn start:web:dev
   ```

2. **Navigate to Task Management**:
   - Open browser at `http://localhost:3030`
   - Go to any team dashboard with tasks

3. **Test Task Assignment Modal**:
   - Click on any task to open the task assignment modal
   - Try to modify any field (title, assignees, status, etc.)
   - Click save/update

4. **Verify Success**:
   - ✅ No Zod validation errors in console
   - ✅ No foreign key constraint errors
   - ✅ Task updates successfully
   - ✅ Changes are persisted and visible

5. **Test Edge Cases**:
   - Tasks with only Direct Employee members
   - Tasks with only OrganizationTeamEmployee members  
   - Tasks with mixed member structures
   - Tasks with no members

### Previous screenshots

**Before Fix - Error State:**
- Console showing Zod validation error: `members.1.employee.email: Required`
- API error: `insert or update on table "task_employee" violates foreign key constraint`
- Task update modal failing to save changes
- User unable to assign tasks or update task properties

https://github.com/user-attachments/assets/b31d2230-d2c5-426b-8792-2b82e080bd22

### Current screenshots

**After Fix - Working State:**
- Task update modal working correctly
- No validation errors in console
- Successful API responses
- Clean normalization logs showing the process
- Task assignments and updates working seamlessly

https://github.com/user-attachments/assets/bc9bf968-1c02-47e8-94d4-6f5b98fda49e

## Related Issues

- Addresses the critical task management bug reported in team Slack discussions
- Resolves user workflow disruption in task assignment functionality
- Fixes data integrity issues with mixed member structures
- Related to Ever Gauzy API contract compliance

## Type of Change

- [x] Bug fix (fixes a problem)
- [ ] New feature (adds functionality)
- [ ] Breaking change (requires changes elsewhere)
- [x] Documentation update

## ✅ Checklist

Please confirm you did the following before asking for review:

- [x] My code follows the project coding style
- [x] I reviewed my own code and added comments where needed
- [x] I tested my changes locally
- [x] I updated or created related documentation if needed
- [x] No new warnings or errors are introduced

## Notes for the Reviewer (Optional)

**Architecture Decisions:**
- **Normalization Approach**: Chose client-side normalization over backend changes to maintain API contract compatibility
- **Functional Programming**: Used `map()` and `filter()` for cleaner, more testable code
- **Backward Compatibility**: Solution works with existing data structures without breaking changes

**Testing Notes:**
- The normalization logic handles all edge cases including null/undefined members
- Zod validation now passes consistently with normalized data
- Performance impact is minimal as normalization only runs during task updates

**Known Considerations:**
- This fix addresses the frontend data structure issue; backend API could potentially be enhanced to handle mixed structures natively in the future
- The solution is defensive and will work even if backend changes are made later

## Summary by CodeRabbit

* **Style**
  * Task modals now enforce a minimum height and block vertical scrolling.
  * Refined spacing, alignment, and consistent icon sizing across task inputs, estimation lists, and member controls.

* **Bug Fixes**
  * Improved handling and display of task members (including team/employee variations), reducing errors when updating or showing members.

### CODE REVIEW

**Comment by evereq** [comment]
@CREDO23 can you review please as first priority?

File: apps/web/core/services/client/api/tasks/task.service.ts [hunk 1]
```diff
@@ -141,6 +141,62 @@ class TaskService extends APIService {
 		}
 	};
 
+	/**
+	 * Determines the type of member based on its structure
+	 */
+	private getMemberType(member: TEmployee): 'DirectEmployee' | 'OrganizationTeamEmployee' {
+		return member?.employeeId ? 'OrganizationTeamEmployee' : 'DirectEmployee';
+	}
```
Comment (CREDO23) [inline apps/web/core/services/client/api/tasks/task.service.ts:149]: Why do we need this @NdekoCode  ?

File: apps/web/core/services/client/api/tasks/task.service.ts [hunk 2]
```diff
@@ -141,6 +141,62 @@ class TaskService extends APIService {
...
+		return members
+			.map((member, index) => {
+				const memberType = this.getMemberType(member);
+
+				return memberType === 'DirectEmployee'
+					? this.normalizeDirectEmployee(member, index)
+					: this.normalizeOrganizationTeamEmployee(member, index);
+			})
+			.filter(Boolean) as TEmployee[];
+	}
```
Comment (CREDO23) [inline apps/web/core/services/client/api/tasks/task.service.ts:199]: Why do we need these adjustments ?

Comment (NdekoCode) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: @CREDO23 These methods solve a critical mixed data structure bug that was breaking task assignments.
`task.members` contains two different structures in the same array:`
```ts
task.members = [
  { userId: "user1", user: {...} },           // Direct Employee ✅
  { employeeId: "emp1", employee: {...} }     // OrganizationTeamEmployee ❌
];
```

The result of that was:
Result:
- Zod error: members.1.employee.email: Required
- FK error: violates foreign key constraint
- Task assignment completely broken

So i used these methods because `getMemberType()` => Identifies structure type`:
```ts
return member?.employeeId ? 'OrganizationTeamEmployee' : 'DirectEmployee';
```

Normalization Methods - Convert to uniform structure:
- DirectEmployee: Already correct, just validate
- OrganizationTeamEmployee: Transform to userId-based structure:
    ```ts
    // Transform this:
    { employeeId: "emp1", employee: { user: {...} } }
    // Into this:
    { userId: "user1", user: {...} }
    ```

So before: we Mixed structures -> Validation fails -> Task assignment broken
And now we Uniformed structures -> Validation passes -> Task assignment works

In short these methods fix a critical bug that made task management unusable. They ensure the API receives consistent data structures as expected.

I could have done this in a single method like in `updateTask`, but that wouldn’t be very scalable or clean. Splitting it into separate methods makes the code easier to maintain, extend, and reason about.

File: apps/web/core/services/client/api/tasks/task.service.ts (hunk 1 above)
Comment (NdekoCode) [reply apps/web/core/services/client/api/tasks/task.service.ts:149]: @CREDO23  check the explanation here https://github.com/ever-co/ever-teams/pull/4035#discussion_r2402774063

File: apps/web/core/services/client/api/tasks/task.service.ts (hunk 2 above)
Comment (NdekoCode) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: I explained all of this in the PR description as well, so feel free to check it out when you have a moment @CREDO23, sir @evereq.

Comment (CREDO23) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: Okay !, but Where does this bug come from ? Why do we have mixed data structure ?

Comment (CREDO23) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: I just want to understand the root problem @NdekoCode !

Comment (CREDO23) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: Is gauzy API that returns mixed structure ? If YES, we should inform Gauzy API guys to fix it so we do not manipulate data in frontend ! What do you think ?? @NdekoCode

Comment (evereq) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: @NdekoCode hm, what if that's a "feature", not a "bug" that API response contains users and employees? We can have tasks assigned to USERS who are NOT employees (e.g. tasks related to interviews assigned to "Candidates" who are NOT employees!). So I assume if API returns such mixed data structure it might be for a reason and you should clarify those things before doing changes....  I am NOT 100% sure about this honestly, i.e. if it's a feature, but I very much in doubt someone will do that by mistake, looks more like it was done intentionally on API side! I think that's one of major issue we have - we see something was done and we don't know why and we assume it was done wrong, while it was possible done for a REASON. Thoughts?

CC: @CREDO23  CC: @rahul-rocket maybe you know something about it?

Comment (CREDO23) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: I will check/test it locally so that I can understand what's going on! @evereq, cc @NdekoCode

Comment (NdekoCode) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: I see your point @evereq. 

My normalization here is more of a defensive approach, just to make sure Ever Teams task assignment doesn’t break no matter what the API returns, and ensures the task assignment continues to work smoothly.  

I’d suggest:  
- I think we can go ahead and merge this PR as assignment issues are unblocked
- After that in long term, if the Gauzy team later (@rahul-rocket) confirms or refactors the API, we can easily align with that.

Comment (evereq) [reply apps/web/core/services/client/api/tasks/task.service.ts:199]: Yes, if your changes not break anything on Ever Teams side, I think we can merge PR :) @CREDO23 please feel free to review more and if need any more changes, let's just create another PR to do more improvements. So that we first fix issues we have and next optimize!
